### ISSUE TITLE
Multi-processing leaking file descriptors

### ISSUE THREAD

**Issue by eliphatfs** [pytorch/pytorch#65198]
## 🐛 Bug
Multiprocessing leaks file descriptors unexpectedly. The issues #11201 and #973 leads us to some workarounds such as to set sharing to file system instead of descriptors, but I think this does not hit the core of the problem and it is also counter-intuitive.

## To Reproduce

1. Run the code (scenario set to 1 or 3).
2. Preferably, run `watch cat /proc/sys/fs/file-nr` in another terminal to watch the number of open file descriptors.

```python
import torch
from torch.utils.data import Dataset, DataLoader
from torch import multiprocessing
# the results are the same using standard MP, which makes me feel strange
# since the error is still raised in pytorch MP
import tqdm
import copy


def process_data(src):
    # Some mock processing logic
    return torch.rand([100])


class TestDataset(Dataset):
    def __len__(self):
        return 200000

    def __getitem__(self, index):
        return torch.rand([100])


if __name__ == "__main__":
    SCENARIO = 3
    if SCENARIO == 1:
        with multiprocessing.Pool(32) as pool:
            data_source = range(200000)  # mock source
            processed_data = list(tqdm.tqdm(pool.imap(process_data, data_source)))
            # Breaks at 65020 iterations here.
            # watch cat /proc/sys/fs/file-nr shows ever-increasing number of FD's
            torch.save(processed_data, 'test_data.pt')
    if SCENARIO == 2:
        with multiprocessing.Pool(32) as pool:
            data_source = range(200000)  # mock source
            processed_data = []
            for d in tqdm.tqdm(pool.imap(process_data, data_source)):
                processed_data.append(copy.deepcopy(d))
                del d
            # This runs successfully
            # watch cat /proc/sys/fs/file-nr shows steady number of FD's
            torch.save(processed_data, 'test_data.pt')
    if SCENARIO == 3:
        # Training model, would like to save data generated for reproducibility
        train_loader = DataLoader(TestDataset(), shuffle=True, num_workers=32)
        data_save = []
        for batch in tqdm.tqdm(train_loader):
            # do something with models
            data_save.append(batch)
        torch.save(data_save, 'test_data.pt')
```

The code produces the error below on the test server:
```
Traceback (most recent call last):
  File "test3.py", line 46, in <module>
    for batch in tqdm.tqdm(train_loader):
  File "/home/***/.local/lib/python3.6/site-packages/tqdm/std.py", line 1185, in __iter__
    for obj in iterable:
  File "/home/***/.local/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 521, in __next__
    data = self._next_data()
  File "/home/***/.local/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 1186, in _next_data
    idx, data = self._get_data()
  File "/home/***/.local/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 1152, in _get_data
    success, data = self._try_get_data()
  File "/home/***/.local/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 990, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/usr/lib/python3.6/multiprocessing/queues.py", line 113, in get
    return _ForkingPickler.loads(res)
  File "/home/***/.local/lib/python3.6/site-packages/torch/multiprocessing/reductions.py", line 294, in rebuild_storage_fd
    storage = cls._new_shared_fd(fd, size)
RuntimeError: unable to mmap 400 bytes from file <filename not specified>: Cannot allocate memory (12)
```

## Expected behavior
Scenario 1 and 3 should work without leaking file descriptors as in scenario 2: if there is a way to gain performance while fixing the leaking FD's then it is excellent; if it is related with performance of sharing tensors, it seems that the copy operation in scenario 2 does not incur much performance overhead, and in real world applications the dataset preprocessing will do heavier work so the overhead will be even smaller compared to the whole process.

If these are not possible, perhaps a warning should be added about possibly leaking FD, or the error message could tell a bit more about the situation. Having a memory allocation error with 750GB RAM available on the server is confusing to me. Wandering randomly on the internet, it takes me days to find the problem.

In addition, on the test server, setting sharing behavior to filesystem in the main block does not fix this problem.

## Environment
 - PyTorch Version (e.g., 1.0): 1.9.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): N/A
 - Python version: 3.6.9
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: P100
 - Any other relevant information:

## Additional context
People may generally use `numpy` for preprocessing. I'm using the `DGL` library with torch backend, so data processing is centered around torch instead.

**Comment by Abdomuller11** [comment]
**k**

**Comment by Abdomuller11** [comment]
*nope*

**Comment by Abdomuller11** [comment]
[goo](http://google.com)

**Comment by rohan-varma** [comment]
This seems more like a data loader issue than distributed training. Looping in @VitalyFedyunin @ejguan.

**Comment by eliphatfs** [comment]
It seems like a problem in `torch.multiprocessing` to me.

**Comment by lw** [comment]
What's the output of `df /dev/shm`?

**Comment by eliphatfs** [comment]
Normally, it looks like
```
Filesystem     1K-blocks  Used Available Use% Mounted on
tmpfs          395610448    40 395610408   1% /dev/shm
```
For scenario 3, the program crashes at `used` ~200,000 used but that's still less than 1%. For `ulimit` it is set to 1,048,576 on my test server. For each process the FD amounts are still fewer than any limit I may find, but it still crashes. Anyways, the leak indeed exists, and should be fixed, I think.

**Comment by lw** [comment]
Are you using Docker (or similar)?

**Comment by eliphatfs** [comment]
No, it is a bare-metal server. Ubuntu 18.04.

**Comment by a-gardner1** [comment]
**tl;dr:** A "leak" does exist, but has an upper bound (that may exceed your system limit). 
As a workaround, reduce the number of tensors per batch. 

The batches obtained in Scenario 3 are shared since they are returned from a multiprocessing dataloader. 
Since you do not deepcopy them but instead retain a reference through the list, the file descriptor cannot be freed.
That is just a problem with the minimal example posted and does not discount the issue you may face in practice.
My team was running into the same issue as Scenario 3 (albeit with deepcopying of each batch) and none of the fixes or workarounds suggested in related issues worked. 
@eliphatfs Do you have a large number of tensors per batch in practice? If

I eventually traced the cause down to [`_MultiprocessingDataloaderIter`](https://github.com/pytorch/pytorch/blob/e5ab0d1013072c26586b369536bccac648843958/torch/utils/data/dataloader.py#L567), namely the [`_next_data` method](https://github.com/pytorch/pytorch/blob/e5ab0d1013072c26586b369536bccac648843958/torch/utils/data/dataloader.py#L1156). 
This method is responsible for retrieving batches from workers and returning them in the order in which the workers were tasked. 
Since the workers are not guaranteed to yield the batches in any particular order due to asynchronous execution, batches retrieved out of order are [placed in a cache](https://github.com/pytorch/pytorch/blob/e5ab0d1013072c26586b369536bccac648843958/torch/utils/data/dataloader.py#L1200). 
As the data from each worker is retrieved from a multiprocessing queue, each tensor contained within it is guaranteed to be shared and have a distinct file descriptor assigned to it (see the [multiprocessing documentation](https://pytorch.org/docs/stable/multiprocessing.html)).
Furthermore, since the data is not deepcopied prior to storage in the cache, the file descriptors will persist until associated task's turn is reached.

The length of the cache is [guaranteed not to exceed the product of the prefetch factor and the number of workers](https://github.com/pytorch/pytorch/blob/e5ab0d1013072c26586b369536bccac648843958/torch/utils/data/dataloader.py#L1206), so one can technically upper bound the number of file descriptors by the number of workers times the prefetch factor.
This cache is also effectively cleared at the end of each epoch (at least under the configurations I tested).
While this is not a leak per se since the descriptors have a definite shelf life and limit, the behavior is certainly unexpected if one is not familiar with the internals of the dataloader and has not knowingly created any shared tensors.

Our workaround was to reduce the number of tensors in a batch by flattening them into a single tensor during collation and extracting them afterwards in the main process.
I am not sure what kind of fix to recommend to `_MultiprocessingDataloaderIter`, if any; deepcopying the data prior to storage in the cache would definitely resolve the issue and avoid the external appearance of a leak.
However, it would definitely introduce some overhead.
I do not think one can return tasks' data out of order received since the order might be important.
Perhaps additional transparency in documentation may help?

On a side note, while this likely depends upon the system if not a feature of some other torch internals, there is not a one-to-one correspondence between the number of shared tensors and the number of file descriptors created with the latter seeming to change only in increments of 224.

**Comment by eliphatfs** [comment]
I don't think that a constant number of 12 is a 'large number of tensors per batch'. And the reproduction script above clearly shows that with only 1 tensor per batch it will still eventually crash the program, so seemingly there isn't such upper bound as O(#tensor * prefetch_factor * num_workers), which given a large constant bound as ~100 is still only ~6400. I see 60000 file descriptors on my server when crashing.

I didn't look carefully at the multiprocessing documentation previously. It really makes some sense -- for particular usage of multi-processing itself, we may want tensors to be shared around processes so they get modified with no synchronization cost.

But for data processing stuff we do not typically need the sharing. Then I suggest:

1. Importing `torch` while using python native `multiprocessing` shall not have this sharing enabled, but uses python-default pickling behavior.
2. An option or context manager to disable sharing.
3. Better error message/warning for file descriptor leaks.

**Comment by a-gardner1** [comment]
Yea, 12 is not a large number of tensors. One of the points I tried to make, though, is that the reproduction script itself contains the leak that causes it to crash.
```python
        data_save = []
        for batch in tqdm.tqdm(train_loader):
            # do something with models
            data_save.append(batch)   # leak
```
If you call `batch.is_shared()`, you will get `True`. The batch will remain in memory as a shared tensor as long as it persists as a member of `data_save`. See [this comment](https://github.com/pytorch/pytorch/issues/11201#issuecomment-486232056) for additional context.

**Comment by eliphatfs** [comment]
Yes, I got that point. But the batch is shared with nobody else, so I think that the clean-up can be made more smart.

**Comment by ejguan** [comment]
>  the batch is shared with nobody else

The tensors reside in the shared memory as multiprocessing involved. So, it's shared across processes.

In your case, you should be able to convert shared tensor to non-shared tensor using `Tensor.narrow_copy`

**Comment by eliphatfs** [comment]
I know I am able to convert it, but I think that the reference tracker could be more smart here.

**Comment by ShenShan123** [comment]
I also encounter this bug when using multiprocessing and a large list containing torch_geometric.data (pyg graphs). Scenario 2 really helps me out. Thanks for your post! And I also found that using a collate() function to compact the large list into one data object is more efficient.
